### PR TITLE
Unpin the `types-setuptools` build dependency

### PR DESCRIPTION
**Pull request by AlexWaygood**
- The latest release of `types-setuptools` started causing type-check errors (and, therefore, mypyc build errors) on the `master` branch: see, e.g. https://github.com/python/mypy/actions/runs/4275505309/jobs/7442902732.
- #14781 addressed some of the new errors, but didn't quite fix the build.
- #14788 then pinned the `types-setuptools` dependency as a temporary stopgap measure.
- This PR now attempts to unpin `types-setuptools` and fix the build errors in a more principled way.

### CODE REVIEW

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Okay, it looks like this PR currently gets us to a place where the mypyc build succeeds, but we then get a ton of failing mypyc tests when the test suite is run :(

**Comment by AlexWaygood** [comment]
I haven't got time to debug the failing tests right now, so here's a stopgap PR pinning `types-setuptools` to avoid the latest release:

- #14788

File: mypyc/build.py [hunk 1]
```diff
@@ -63,13 +63,14 @@ def get_extension() -> type[Extension]:
     # We can work with either setuptools or distutils, and pick setuptools
     # if it has been imported.
     use_setuptools = "setuptools" in sys.modules
+    extension_class: type[Extension]
 
     if not use_setuptools:
-        from distutils.core import Extension
+        from distutils.core import Extension as extension_class
```
Comment (JelleZijlstra) [inline mypyc/build.py:69]:
```suggestion
        import distutils.core
        extension_class = distutils.core.Extension
```
Just a hunch

Comment (AlexWaygood) [reply mypyc/build.py]: Trying it out now...

Comment (AlexWaygood) [reply mypyc/build.py]: And it worked!

File: pyproject.toml [hunk 2]
```diff
@@ -12,7 +12,7 @@ requires = [
     "tomli>=1.1.0; python_version<'3.11'",
     # the following is from build-requirements.txt
     "types-psutil",
-    "types-setuptools<67.4.0.2",  # TODO: fix build when using the latest version of types-setuptools
+    "types-setuptools",
```
Comment (hauntsaninja) [inline pyproject.toml:15]: Does this need a lower bound?

Comment (AlexWaygood) [reply pyproject.toml:15]: I don't _think_ so -- https://github.com/python/mypy/pull/14781 had successful CI runs with `types-setuptools==67.4.0.1`, e.g. https://github.com/python/mypy/actions/runs/4272132207/jobs/7437069696. But I can add one anyway to be sure, if you like -- it probably wouldn't hurt

**Pull request merged**